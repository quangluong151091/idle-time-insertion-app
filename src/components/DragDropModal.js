import React from 'react';
import { Modal } from 'react-bootstrap';
import { Button } from '@progress/kendo-react-buttons';
import DatePicker from "react-datepicker";
import "react-datepicker/dist/react-datepicker.css";

class DragDropModal extends React.Component {
    constructor(props, context) {
        super(props, context);

        this.handleShow = this.handleShow.bind(this);
        this.handleClose = this.handleClose.bind(this);

        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleCancel = this.handleCancel.bind(this);
        this.handleChange = this.handleChange.bind(this);
        this.handleDateChange = this.handleDateChange.bind(this);

        this.state = {
            show: this.props.show,
            startDate: new Date(),
            item: ""
        };
    }
    componentWillReceiveProps(nextProps) {
        if (nextProps.show !== this.state.show) {
            this.setState({ show: nextProps.show });
        }
    }
    handleClose() {

        this.setState({ show: false });
        this.props.closeModal();
    }

    handleShow() {
        this.setState({ show: true });
    }
    handleChange(e) {
        this.setState({ item: e.target.value })
    }
    handleDateChange(date) {
        this.setState({
            startDate: date
        });
    }
    handleCancel() {

        this.setState({ item: "" });
        this.setState({ show: false });
        this.props.closeModal();
    }
    formatDateStr(d) {
        d = d.toString();
        d = d.split("G")[0]
    
        return d;
    }
    handleSubmit() {
        this.setState({ show: false });

        var itemsArray = JSON.parse(localStorage.getItem(this.props.name));
        var newItem = {
            "author": "Demo user",
            "comment": this.state.item,
            "editable": true,
            "noteId": "Note",
            "type": 'dot',
            "reason": this.props.part,
            "content": "<img src='/pics/stickers/" + this.props.item + ".png' height='42' width='42' />"
        };
        newItem.start = new Date(this.state.startDate);
        newItem.startStr = this.formatDateStr(new Date(this.state.startDate));
        newItem.dateStr = this.formatDateStr(new Date());
        itemsArray.push(newItem);
        localStorage.setItem(this.props.name, JSON.stringify(itemsArray));

        this.props.closeModal();
    }
    render() {
        return (
            <div>
                <Modal show={this.state.show} onHide={this.handleClose}>
                    <Modal.Header closeButton>
                        <Modal.Title><img src='../../../img\/note.png' alt="note" /> Create new note</Modal.Title>
                    </Modal.Header>
                    <Modal.Body>
                        <div className="panel panel-default">
                            <div className="panel-body">
                                <span className="font-weight-bold"><b>Time of error: </b></span>
                                <DatePicker
                                    selected={this.state.startDate}
                                    onChange={this.handleDateChange}
                                    showTimeSelect
                                    timeFormat="HH:mm"
                                    timeIntervals={1}
                                    dateFormat="dd/MM/yyyy HH:mm"
                                    timeCaption="time"
                                />
                            </div>
                        </div>
                        <div id="detail">
                            <span className="font-weight-bold">Additional description: </span>
                            <textarea onChange={this.handleChange} className="cw-modal-text-area" />
                        </div>
                    </Modal.Body>
                    <Modal.Footer>
                        <Button id="cancel" onClick={this.handleCancel}>Cancel</Button>
                        <Button id="submit" primary={true} onClick={this.handleSubmit}>Submit</Button>
                    </Modal.Footer>
                </Modal>
            </div>
        );
    }
}

export default DragDropModal;
