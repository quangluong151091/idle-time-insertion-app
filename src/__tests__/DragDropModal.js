import React from 'react';
import { shallow } from 'enzyme';

import DragDropModal from '../components/DragDropModal';

describe('DragDropModal', () => {
  it('should render correctly', () => {

    const component = shallow(<DragDropModal />);
    
    component.setState({'startDate': new Date("12/12/2018 10:00")});
    expect(component).toMatchSnapshot();
  });
  it('should be able to open', () => {

    const component = shallow(<DragDropModal show={true}/>);
    
    expect(component.state("show")).toEqual(true);
  });
  it('should change textarea input', () => {

    const component = shallow(<DragDropModal/>);
    
    component.find('.cw-modal-text-area').simulate('change', {'target': {'value': '123'}});
    expect(component.state("item")).toEqual('123');
  });
  it('should change date input', () => {

    const component = shallow(<DragDropModal/>);
    
    component.find('DatePicker').simulate('change', new Date("12/12/2018 12:00"));
    expect(component.state("startDate")).toEqual(new Date("12/12/2018 12:00"));
  });
  it('cancel form when clicked', () => {
    const mock = jest.fn();
    const component = shallow(<DragDropModal closeModal={mock}/>);
    const button = component.find('#cancel');
    button.simulate('click');
    expect(component.state("item")).toEqual('');
    expect(component.state("show")).toEqual(false);
  });
  it('submit form when clicked', () => {
    const mock = jest.fn();
    localStorage.setItem("29484001", "[{}]");
    const component = shallow(<DragDropModal closeModal={mock} part="29484" item="enduser" name="29484001"/>);
    const button = component.find('#submit');
    button.simulate('click');

    expect(component.state("show")).toEqual(false);
    expect(localStorage.length).toEqual(1);
  });
});