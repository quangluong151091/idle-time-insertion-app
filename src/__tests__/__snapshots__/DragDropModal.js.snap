// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DragDropModal should render correctly 1`] = `
<div>
  <Modal
    animation={true}
    autoFocus={true}
    backdrop={true}
    bsClass="modal"
    dialogComponentClass={[Function]}
    enforceFocus={true}
    keyboard={true}
    manager={
      ModalManager {
        "add": [Function],
        "containers": Array [],
        "data": Array [],
        "handleContainerOverflow": true,
        "hideSiblingNodes": true,
        "isTopModal": [Function],
        "modals": Array [],
        "remove": [Function],
      }
    }
    onHide={[Function]}
    renderBackdrop={[Function]}
    restoreFocus={true}
    show={false}
  >
    <ModalHeader
      bsClass="modal-header"
      closeButton={true}
      closeLabel="Close"
    >
      <ModalTitle
        bsClass="modal-title"
        componentClass="h4"
      >
        <img
          alt="note"
          src="../../../img\\\\/note.png"
        />
         Create new note
      </ModalTitle>
    </ModalHeader>
    <ModalBody
      bsClass="modal-body"
      componentClass="div"
    >
      <div
        className="panel panel-default"
      >
        <div
          className="panel-body"
        >
          <span
            className="font-weight-bold"
          >
            <b>
              Time of error: 
            </b>
          </span>
          <DatePicker
            allowSameDay={false}
            dateFormat="dd/MM/yyyy HH:mm"
            dateFormatCalendar="LLLL yyyy"
            disabled={false}
            disabledKeyboardNavigation={false}
            dropdownMode="scroll"
            monthsShown={1}
            nextMonthButtonLabel="Next month"
            onBlur={[Function]}
            onChange={[Function]}
            onClickOutside={[Function]}
            onFocus={[Function]}
            onInputClick={[Function]}
            onInputError={[Function]}
            onKeyDown={[Function]}
            onMonthChange={[Function]}
            onSelect={[Function]}
            onYearChange={[Function]}
            preventOpenOnFocus={false}
            previousMonthButtonLabel="Previous Month"
            readOnly={false}
            renderDayContents={[Function]}
            selected={2018-12-12T08:00:00.000Z}
            shouldCloseOnSelect={true}
            showTimeSelect={true}
            timeCaption="time"
            timeFormat="HH:mm"
            timeIntervals={1}
            withPortal={false}
          />
        </div>
      </div>
      <div
        id="detail"
      >
        <span
          className="font-weight-bold"
        >
          Additional description: 
        </span>
        <textarea
          className="cw-modal-text-area"
          onChange={[Function]}
        />
      </div>
    </ModalBody>
    <ModalFooter
      bsClass="modal-footer"
      componentClass="div"
    >
      <Button
        id="cancel"
        look="default"
        onClick={[Function]}
        primary={false}
        selected={false}
        togglable={false}
      >
        Cancel
      </Button>
      <Button
        id="submit"
        look="default"
        onClick={[Function]}
        primary={true}
        selected={false}
        togglable={false}
      >
        Submit
      </Button>
    </ModalFooter>
  </Modal>
</div>
`;
